sympy/matrices/expressions/applyfunc.py

```diff
@@ -72,19 +72,28 @@ def func(self):
         # This strange construction is required by the assumptions:
         # (.func needs to be a class)
 
-        class ElementwiseApplyFunction2(ElementwiseApplyFunction):
-            def __new__(obj, expr):
+        class _(ElementwiseApplyFunction):
+            def __new__(cls, expr):
                 return ElementwiseApplyFunction(self.function, expr)
```

Comment:
I know this isn't being added in this PR but this construction seems like a bad idea. The function should be part of the args rather than messing with func like this.

Reply:
Well, this way you keep the same parameters as in the function definition. Methods applying to functions should also apply here.

Reply:
Can you give an example of why this is needed? Why can't the class be (basically):
```
class ElementWiseApplyFunction(MatrixExpr):
    def __new__(cls, function, expr):
        # Pass function as part of args
        obj = MatrixExpr.__new__(cls, function, expr)
```
A similar construction using classes created on the fly is used for `UndefinedFunction` but I think that should be changed as well.

Reply:
A generic Python function may not be sympifiable. That's why the argument is hidden.

Reply:
A generic Python function is turned into a Lambda:
https://github.com/sympy/sympy/blob/fcf84d99dd243b9921eaebe19281b692feb7869a/sympy/matrices/expressions/applyfunc.py#L46-L51